# Bacteria and archaea phenotypic traits

The coded workflow corresponding with the data descriptor: Madin et al. (XXXX) A global synthesis of bacterial and archaeal phenotypic trait and environment data. Scientific Data XXXX 

### Current datasets included in merger workflow

1. bacdive-microa.csv
1. bergeys.csv
1. campedelli.csv
1. corkrey.csv
1. edwards.csv
1. engqvist.csv
1. fierer.csv
1. genbank.csv
1. gold.csv
1. jemma-refseq.csv
1. kegg.csv
1. kremer.csv
1. masonmm.csv
1. mediadb.csv
1. methanogen.csv
1. microbe-directory.csv
1. nielsensl.csv
1. pasteur.csv
1. patric.csv
1. prochlorococcus.csv
1. protraits.csv
1. rrndb.csv
1. silva.csv

### Overview of project files

Directory | Content
--- | ---
`data/` | Contains all the raw data. This directory is read only (i.e., do not change these files!)
`data/conversion_tables/` | Contains tables for mapping traits among datasets to ensure standard naming and values. This directory also contains the data_corrections file for changes to records with errors.  
`data/raw/` | Contains the raw data file for each of the datasets in the merger. Each raw data sub-directory contains a README file that explains where the dataset came from and notes about how it was processed.
`data/taxonomy/` | Contains the NCBI taxonomy database dump. 
`R/` | All R code for preparing and merging/condensing the final datasets, including files containing global functions, loading necessary R packages, and global settings (outlines below). Data preparation files are in the `preparation` directory. The main merge files are `condense_species.R` and `condense_traits.R`.
`R/preparation` | Contains the individual dataset preparation scripts. 
`R/taxonomy` | Contains the NCBI taxonomy and taxonomic mapping preparation scripts. 
`R/web_extraction` | Contains script for utilizing APIs for extracting data from web-based datasets. 
`output/` | For all R code output that is used for analysis or during the merge and preparation process. Everything in this folder can be regenerated by the `R` code.
`output/prepared_data/` | Contains the processed 'cleaned up' version of all included data frames - these are generated by the individual clean scripts for each dataset and are loaded into the merger code
`output/prepared_references/` | Contains the processed reference table.
`output/taxonomy/` | Contains pre-processed taxonomy maps and the main NCBI mapping table with other taxonomic levels. 
`.gitignore` | What should be ignored by git versioning; also won't appear on Github. This currently contains important files that are too large for Github, like the NCBI taxonomy and the Bergey's PDFs.
`README.md` | This is what you're reading. There are other "README" files through the file system, including the data directory for information about datasets, etc.
`workflow.R` | The final analysis code that loads and prepares each raw dataset and then runs the merge and condensations. 
`bacteria_archaea_traits.Rproj` | You double-click on this when you've cloned the GitHub project to your computer, and the project is opened in RStudio. Otherwise, this file that can be ignored.

### Large datasets not included in Github

There are two datasets too large to be kept at GitHub, and these are automatically downloaded from Figshare when the `workflow.R` is first run. If you experience problems, these files need to be manually downloaded from FigShare and placed in the project directory as outlined here.

File | Link | Where to put
--- | --- | ---
`genome_metadata.txt` | ftp://ftp.patricbrc.org/RELEASE_NOTES/ | `data/raw/patric/`
`taxonomy_names.csv` | https://figshare.com/s/ab40d2a35266d729698c | `output/taxonomy/`

### Overview of settings file

The `settings.R` file in the `R` directory is where the different scripts are told which data sets to include in the merger and how to handle different traits and whether they will need to be translated. This file will need to be updated if new traits are added to the merger by inserting the respective column name into the appropriate vector variable (see description below).

Any varaible that is required in the preparation code is named in capital and with the word "CONSTANT_" in front so as to avoid being accidentally overwritten.

Variable | Description
--- | ---
`CONSTANT_DATA_PATH` | This is the path to the folder containing all prepared data for merging
`CONSTANT_EXCLUDED_DATASETS` | This is a vector of the names of data sets that should NOT be included in the following merger
`CONSTANT_CATEGORICAL_DATA_COLUMNS` | Vector with the name of columns holding categorical trait data
`CONSTANT_CONTINOUS_DATA_COLUMNS` | Vector with the name of columns holding continuous (integers/doubles/floats) trait data
`CONSTANT_OTHER_COLUMNS` | Vector with the name of columns holding non-data information such as phylogenetic categories
`CONSTANT_ALL_DATA_COLUMNS` | Vector holding all data columns
`CONSTANT_FINAL_COLUMNS` | Vector holding all columns to be included in final output
`CONSTANT_DATA_FOR_RENAMING` | Vector holding the name of columns where the data will need to be re-named according to respective lookup tables
`CONSTANT_SPECIAL_CATEGORICAL_TRAITS` | Vector holding holding column names of traits that should NOT be condensed using the standard condensation code
`CONSTANT_GENERAL_CATEGORICAL_PROCESSING` | Vector holding column names of traits that SHOULD be condensed using the standard condensation code
`CONSTANT_DOMINANT_TRAIT_PROPORTION` | The proportion (out of 100) that a trait value must occupy of a population in order to be selected as the appropriate representation of that species
`CONSTANT_FILL_GTDB_WITH_NCBI` | TRUE/FALSE Should the NCBI phylogeny should be used as space filler in the GTDB phylogeny
`CONSTANT_DOMINANT_TRAIT_PRIORITISE` | Sets whether the "max" or "min" stringent value (i.e. "obligate aerobe"" is more stringent and "aerobe") for a given trait should be chosen in case of a tie between two values
`CONSTANT_GROWTH_RATE_ADJUSTMENT_FINAL_TMP` | Temperature for calculating temperature standardised growth rates using Q10
`CONSTANT_GROWTH_RATE_ADJUSTMENT_Q10` | Q10 value for standardised growth rate calculations

### Merged and condensed files

One setting is kept in `workflow.R`. CONSTANT_BASE_PHYLOGENY is for choosing the taxonomy to be used for condensation ("NCBI" or Genome Taxonomy Database "GTDB")

The `condensed_traits` is all the datasets combined and condensed into one row per trait and combined with full phylogeny based on the NCBI taxonomy map. 

The `condensed_species` is produced from the `condensed_traits` where data has been condensed to one row per species as per the NCBI taxonomy.

- [condensed_traits.csv](output/condensed_traits.csv)
- [condensed_species.csv](output/condensed_species.csv)

### Run scripts

R version [>3.5]

If you want to run the full merger including the preparation of each raw dataset, then you only need to run the `workflow.R` script in the root folder

`source("workflow.R")`

If you have updated a preparation (clean) script for an individual dataset or added a new dataset, you only need to run the particular clean-up script for that data set and introduce the new data into the merger by re-running (in that order):

1. `condense_traits.R`
2. `condense_species.R`
